Implement userdata export by creating a FAT16 filesystem

megari opened the pull request:
This PR implements exporting userdata in an easier-to-use manner, by creating a 16 MiB FAT16 volume on the SD card, with the userdata dump as a file in the root directory.

The code could probably be made smaller by taking additional liberties with regard to the FAT16 specification. Currently, the only violation I know of is that only the first sector of the root directory is zeroed out, apart from the directory entry for the userdata dump.

megari committed:
Userdata export: export settings on a FAT16 filesystem.

megari committed:
Userdata export: Zero out whole FAT area and handle SD write errors.

megari committed:
Userdata export: Fix regression in generating FAT in multiple steps.

megari committed:
Userdata export: Fix remaining regressions in FAT generation.

megari commented:
The delta to the previous userdata export PR works out to 546 or 548 bytes. Not a whole lot, but can probably be reduced further.

Previous PR post-build statistics:
```
   text	   data	    bss	    dec	    hex	filename
  31956	   2936	   2652	  37544	   92a8	sys_controller.elf
riscv32-unknown-elf-objcopy --change-addresses -0x10000 -O binary --gap-fill 0 sys_controller.elf mem_init/sys_onchip_memory2_0.bin
../../tools/bin2hex 4 mem_init/sys_onchip_memory2_0.bin mem_init/sys_onchip_memory2_0.hex
size: 34892 bytes
```

The current post-build statistics:
```
   text	   data	    bss	    dec	    hex	filename
  32502	   2936	   2652	  38090	   94ca	sys_controller.elf
riscv32-unknown-elf-objcopy --change-addresses -0x10000 -O binary --gap-fill 0 sys_controller.elf mem_init/sys_onchip_memory2_0.bin
../../tools/bin2hex 4 mem_init/sys_onchip_memory2_0.bin mem_init/sys_onchip_memory2_0.hex
size: 35440 bytes
```

alanbork commented:
personally I'd rather save the space for features than file export which I have little expectation to use...  but I don't have a sense on how tight space is at this point.

megari committed:
Userdata export (FAT16): Optimize code size.

Reorganize the and optimize the code to avoid redundant code and thus
reduce code size, while making sure to zero-initialize the used areas
on the SD card. These are the "clean" part of this change.

The "dirty" part:
Optimize the FAT16 export code further by introducing small deviations
from the FAT16 specification. These deviations should not be harmful
at all, unless the SD card is used for something requiring the jump
instruction and/or boot code in the boot sector to be valid. This is
typically only required when booting off the filesystem.

After these changes, a considerable reduction in code size can be
observed for sys_controller.elf and sys_onchip_memory2_0.bin:

sys_controller.elf:
   text	   data	    bss	    dec	    hex	filename
  32392	   2936	   2652	  37980	   945c	sys_controller.elf

sys_onchip_memory2_0.bin:
size: 35328 bytes

This reduces the cost of the FAT16 export feature to 446 bytes.

megari commented:
> personally I'd rather save the space for features than file export which I have little expectation to use... but I don't have a sense on how tight space is at this point.

After some optimization, this feature now requires 446 bytes. This is unlikely to go substantially lower. This leaves a margin of about 2 kB. This may be enough for a few small features in addition to any bugfixes. Let's see how this fits in the roadmap.

megari committed:
Userdata export: provide Japanese translations for prompt.

Checked with a native speaker working in the software industry.

megari committed:
Userdata export: Translate status and error messages to Japanese.

megari committed:
menu.c: Translate name of Settings opt menu to Japanese.

megari committed:
menu.c: Fix a typo in a Japanese translation.

This looks like a simple case of mistyping the ヲ particle, denoting
the object in the menu title.
